book/src/developers/designs/cryptography-key-domains.md

```diff
@@ -0,0 +1,88 @@
+# Cryptography Key Domains
+
+Within Kanidm we have to manage a number of private keys with various cryptograhpic purposes. In the
+current design, we have evolved where for each purposes keys are managed in unique ways. However we
+need to improve this for a number reasons including shared keys for Oauth2 domains and a future
+integration with PKCS11.
+
+## Current System
+
+In the current system we store keys in database records in a variety of bespoke ways for different
+uses. Often this means that keys are associated to the object they are integrated with, such as the
+oauth2 client configuration. However this limits us since we may have users that wish to combine
+multiple oauth2 clients into a single security domain, where access tokens may be exchanged between
+clients for forwarded authentication.
+
+Another example is api-tokens for service accounts. In the past we associated a private key to each
+service-account, however this causes issues with key management. We want to combine all of these
+keys but preserve existing api-token signatures. Currently we have no mechanism to do this.
+
+## Use Cases
+
+- Multiple OAuth2 clients wish to be able to access a shared resource server. `access_tokens` issued
+  to client a or b should be accepted by the resource server.
+- Support key-rotation within a key-domain, where former public keys are retained in a valid state,
+  or can be revoked.
+- Replication Coordinator role needs a way to issue keys that only work within the replication
+  coordinator scope.
+- Migration of key-domains and private keys into PKCS11 modules, or creation of new keys in PKCS11
+  to replace former keys.
+
+## Design
+
+Keys will be moved to dedicated key-objects in the database. Each key-object can have many
+associated keys of various cryptographic types.
+
+For each key present, they have a key-id related to the public keys. This allows lookup based on
```

Comment:
Need multiple key-ids
Need to check pkcs11 key id format as well as jwt. Likely binary, probably not fixed size. Not sure how to index that.

Reply:
I could see the data being like this:

```
struct/db table KeyObject
- key_object_id: uuid

struct/db table KeyKey
- key_id: uuid
- key_object: uuid
- public_key: String? (at least in the db 😄 )

enum KeyIdType:
  Jwt (implies String)
  PKCS11 (implies Base64 of keyid)

KeyId
- keytype: KeyIdType
- key_id: uuid
- value: String
```

```mermaid
erDiagram
    KeyObject ||--o{ KeyKey : key_object_id
    KeyKey ||--|{ KeyId : key_id
```

Reply:
Yeah, I was thinking similar. I need to do another pkcs11 code test to check stuff meshes "approximately".